modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteHashIndexSpool.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.core.Spool;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.ignite.internal.processors.query.calcite.metadata.cost.IgniteCost;
+import org.apache.ignite.internal.processors.query.calcite.metadata.cost.IgniteCostFactory;
+import org.apache.ignite.internal.processors.query.calcite.util.RexUtils;
+import org.apache.ignite.internal.util.typedef.F;
+
+/**
+ * Relational operator that returns the sorted contents of a table
+ * and allow to lookup rows by specified bounds.
+ */
+public class IgniteHashIndexSpool extends Spool implements IgniteRel {
+    /** Search row. */
+    private final List<RexNode> searchRow;
+
+    /** Search row. */
+    private final ImmutableBitSet keys;
+
+    /** Condition (used to calculate selectivity). */
+    private final RexNode cond;
+
+    /** */
+    public IgniteHashIndexSpool(
+        RelOptCluster cluster,
+        RelTraitSet traits,
+        RelNode input,
+        List<RexNode> searchRow,
+        RexNode cond
+    ) {
+        super(cluster, traits, input, Type.LAZY, Type.EAGER);
+
+        assert !F.isEmpty(searchRow);
+
+        this.searchRow = searchRow;
+        this.cond = cond;
+
+        keys = ImmutableBitSet.of(RexUtils.notNullKeys(searchRow));
+    }
+
+    /**
+     * Constructor used for deserialization.
+     *
+     * @param input Serialized representation.
+     */
+    public IgniteHashIndexSpool(RelInput input) {
+        this(input.getCluster(),
+            input.getTraitSet().replace(IgniteConvention.INSTANCE),
+            input.getInputs().get(0),
+            input.getExpressionList("searchRow"),
+            null
+        );
+    }
+
+    /** {@inheritDoc} */
+    @Override public <T> T accept(IgniteRelVisitor<T> visitor) {
+        return visitor.visit(this);
+    }
+
+    /** */
+    @Override public IgniteRel clone(RelOptCluster cluster, List<IgniteRel> inputs) {
+        return new IgniteHashIndexSpool(cluster, getTraitSet(), inputs.get(0), searchRow, cond);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected Spool copy(RelTraitSet traitSet, RelNode input, Type readType, Type writeType) {
+        return new IgniteHashIndexSpool(getCluster(), traitSet, input, searchRow, cond);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean isEnforcer() {
+        return true;
+    }
+
+    /** */
+    @Override public RelWriter explainTerms(RelWriter pw) {
+        RelWriter writer = super.explainTerms(pw);
+
+        return writer.item("searchRow", searchRow);
+    }
+
+    /** {@inheritDoc} */
+    @Override public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+        double rowCnt = mq.getRowCount(getInput());
+        double bytesPerRow = (getRowType().getFieldCount() - keys.cardinality()) * IgniteCost.AVERAGE_FIELD_SIZE;
```

Comment:
why did you this subtraction? the whole row is stored within the index

Reply:
Right. I think about reduce row but have not implemented yet.